sygnal/firebasepushkin.py

```diff
@@ -0,0 +1,263 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+from typing import Dict, Optional
+
+import attr
+from firebase_admin import credentials, initialize_app, messaging
+from prometheus_client import Histogram
+from twisted.internet.defer import Deferred
+from twisted.python.threadpool import ThreadPool
+
+from .exceptions import PushkinSetupException
+from .notifications import Pushkin
+
+SEND_TIME_HISTOGRAM = Histogram(
+    "sygnal_fcm_request_time", "Time taken to send HTTP request"
+)
+
+MAX_TRIES = 3
+RETRY_DELAY_BASE = 10
+MAX_BYTES_PER_FIELD = 1024
+DEFAULT_MAX_CONNECTIONS = 20
+
+logger = logging.getLogger(__name__)
+
+
+@attr.s
+class FirebaseConfig(object):
+    credentials = attr.ib()
+    max_connections = attr.ib(default=20)
+    message_types = attr.ib(default=attr.Factory(dict), type=Dict[str, str])
+
+
+class FirebasePushkin(Pushkin):
+    def __init__(self, name, sygnal, config):
+        super(FirebasePushkin, self).__init__(name, sygnal, config)
+
+        self.db = sygnal.database
+        self.reactor = sygnal.reactor
+        self.config = FirebaseConfig(
+            **{x: y for x, y in self.cfg.items() if x != "type"}
+        )
+
+        credential_path = self.config.credentials
+        if not credential_path:
+            raise PushkinSetupException("No Credential path set in config")
+
+        cred = credentials.Certificate(credential_path)
+
+        self._pool = ThreadPool(maxthreads=self.config.max_connections)
+        self._pool.start()
+
+        self._app = initialize_app(cred, name="app")
+
+    async def dispatch_notification(self, n, device, context):
+
+        pushkeys = [
+            device.pushkey for device in n.devices if device.app_id == self.name
+        ]
+
+        failed = []
+        data = self.build_message(n)
+
+        logger.debug("Type: %s", data["type"])
+
+        if (
+            data["type"] != "m.room.message"
+            or data["content"]["msgtype"] not in self.config.message_types
+        ):
+            return failed
```

Comment:
So we're not sending push for non-message events? What about invites and things?